platform/linux-generic/odp_packet.c

```diff
@@ -1444,6 +1710,109 @@ int odp_packet_split(odp_packet_t *pkt, uint32_t len, odp_packet_t *tail)
 }
 
 /*
+ * References
+ */
+
+static inline void packet_ref(odp_packet_hdr_t *pkt_hdr)
+{
+	uint32_t i;
+	odp_packet_hdr_t *hdr;
+
+	do {
+		for (i = 0; i < pkt_hdr->buf_hdr.segcount; i++) {
+			hdr = pkt_hdr->buf_hdr.seg[i].hdr;
+			packet_ref_inc(hdr);
+		}
+
+		pkt_hdr = pkt_hdr->ref_hdr;
+	} while (pkt_hdr);
+}
+
+static inline odp_packet_t packet_splice(odp_packet_hdr_t *pkt_hdr,
+					 uint32_t offset,
+					 odp_packet_hdr_t *ref_hdr)
+{
+	/* Catch attempted references to stale handles in debug builds */
+	ODP_ASSERT(packet_ref_count(pkt_hdr) > 0);
+
+	/* Splicing is from the last section of src pkt */
+	while (ref_hdr->ref_hdr)
+		ref_hdr = ref_hdr->ref_hdr;
+
+	/* Find section where splice begins */
+	while (offset >= pkt_hdr->frame_len && pkt_hdr->ref_hdr) {
+		offset   -= (pkt_hdr->frame_len - pkt_hdr->ref_offset);
+		offset   += (pkt_hdr->ref_hdr->frame_len - pkt_hdr->ref_len);
```

Comment:
Here can suggest to use macros to write code in self explanation, like below (which seems not equal, but just express the opinion of writing code in self explanation):

reference_length = packet->frame_len;
while (offset > reference_length && has_next_reference(packet)) {
    offset -= reference_length;
    reference_length = next_reference_length(packet);
    packet = next_reference(packet);
}

#define next_reference(packet) packet->ref_hdr
#define has_next_reference(packet) (packet->ref_hdr != NULL)

#define next_reference_length(packet) \
    captured_reference_length(packet) + reference_length_compensation(packet)

/* the reference length captured when the reference was created */
#define captured_reference_length(packet) (packet->ref_len - packet->ref_offset)

/* the reference length compensation due to frame manipulation after reference creation */
#define reference_length_compensation(packet) (packet->ref_hdr->frame_len - packet->ref_len)